src/Microsoft.AspNet.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -12,27 +13,44 @@ namespace Microsoft.AspNet.Server.Kestrel.Filter
 {
     public class StreamSocketOutput : ISocketOutput
     {
+        private static readonly byte[] _endChunkBytes = Encoding.ASCII.GetBytes("\r\n");
         private static readonly byte[] _nullBuffer = new byte[0];
 
         private readonly Stream _outputStream;
         private readonly MemoryPool2 _memory;
         private MemoryPoolBlock2 _producingBlock;
 
+        private object _writeLock = new object();
+
         public StreamSocketOutput(Stream outputStream, MemoryPool2 memory)
         {
             _outputStream = outputStream;
             _memory = memory;
         }
 
-        void ISocketOutput.Write(ArraySegment<byte> buffer, bool immediate)
+        public void Write(ArraySegment<byte> buffer, bool immediate, bool chunk)
         {
-            _outputStream.Write(buffer.Array ?? _nullBuffer, buffer.Offset, buffer.Count);
+            lock (_writeLock)
+            {
+                if (chunk && buffer.Array != null)
+                {
+                    var beginChunkBytes = ChunkWriter.BeginChunkBytes(buffer.Count);
+                    _outputStream.Write(beginChunkBytes.Array, beginChunkBytes.Offset, beginChunkBytes.Count);
+                }
+
+                _outputStream.Write(buffer.Array ?? _nullBuffer, buffer.Offset, buffer.Count);
+
+                if (chunk && buffer.Array != null)
+                {
+                    _outputStream.Write(_endChunkBytes, 0, _endChunkBytes.Length);
+                }
+            }
         }
 
-        Task ISocketOutput.WriteAsync(ArraySegment<byte> buffer, bool immediate, CancellationToken cancellationToken)
+        public Task WriteAsync(ArraySegment<byte> buffer, bool immediate, bool chunk, CancellationToken cancellationToken)
         {
             // TODO: Use _outputStream.WriteAsync
```

Comment:
K, so tests fail changing this to async. Will dig in and do a different PR if I get a resolution.